gobblin-config-management/gobblin-config-core/src/main/java/gobblin/config/common/impl/InMemoryValueInspector.java

```diff
@@ -0,0 +1,87 @@
+package gobblin.config.common.impl;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.WeakHashMap;
+
+import com.typesafe.config.Config;
+
+import gobblin.config.store.api.ConfigKeyPath;
+
+/**
+ * InMemoryValueInspector provide the caching layer for getting the {@com.typesafe.config.Config} from {@ConfigStore}
+ * 
+ * @author mitu
+ *
+ */
+public class InMemoryValueInspector implements ConfigStoreValueInspector{
+
+  private final ConfigStoreValueInspector valueFallback;
+
+  private final Map<ConfigKeyPath, Config> ownConfigMap;
+  private final Map<ConfigKeyPath, Config> recursiveConfigMap;
+
+  /**
+   * 
+   * @param valueFallback - the fall back {@ConfigStoreValueInspector} which used to get the raw {@com.typesafe.config.Config} 
+   * @param useStrongRef  - if true, use Strong reference in cache, else, use Weak reference in cache
+   */
+  public InMemoryValueInspector (ConfigStoreValueInspector valueFallback, boolean useStrongRef){
+    this.valueFallback = valueFallback;
+
+    if(useStrongRef){
+      this.ownConfigMap = new HashMap<>();
```

Comment:
You want to use at least a `ConcurrentHashMap` to ensure thread safety. However, I would suggest using Guava cache instead.